File: project/jimmer-sql/src/test/java/org/babyfish/jimmer/sql/dto/PolymorphicDtoInputTest.java
package org.babyfish.jimmer.sql.dto;

import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import com.fasterxml.jackson.annotation.JsonTypeName;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.babyfish.jimmer.runtime.ImmutableSpi;
import org.babyfish.jimmer.sql.model.inheritance.enumdiscriminator.ClientType;
import org.babyfish.jimmer.sql.model.inheritance.enumdiscriminator.EnumClient;
import org.babyfish.jimmer.sql.model.inheritance.enumdiscriminator.EnumPerson;
import org.babyfish.jimmer.sql.model.inheritance.enumdiscriminator.dto.EnumClientDiscriminatorInput;
import org.babyfish.jimmer.sql.model.inheritance.joinedtable.instantiable.dto.InstantiableClientDefaultInput;
import org.babyfish.jimmer.sql.model.inheritance.singletable.Client;
import org.babyfish.jimmer.sql.model.inheritance.singletable.Organization;
import org.babyfish.jimmer.sql.model.inheritance.singletable.Person;
import org.babyfish.jimmer.sql.model.inheritance.singletable.dto.*;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

public class PolymorphicDtoInputTest {

    private final ObjectMapper mapper = new ObjectMapper();

    @Test
    public void testImplicitDefaultInputCreatesRootEntityShape() {
        ClientPatchInput.Default input = new ClientPatchInput.Default();
        input.setId(10L);
        input.setName("Base patch");

        Client entity = input.toEntity();

        Assertions.assertEquals(Client.class, ((ImmutableSpi) entity).__type().getJavaClass());
        Assertions.assertEquals(10L, entity.id());
        Assertions.assertEquals("Base patch", entity.name());
    }

    @Test
    public void testExplicitBranchInputCreatesDerivedTypeEntityShape() {
        ClientPatchInput.Organization input = new ClientPatchInput.Organization();
        input.setId(11L);
        input.setName("Org patch");
        input.setTaxCode("T-11");

        Organization entity = input.toEntity();

        Assertions.assertEquals(Organization.class, ((ImmutableSpi) entity).__type().getJavaClass());
        Assertions.assertEquals(11L, entity.id());
        Assertions.assertEquals("Org patch", entity.name());
        Assertions.assertEquals("T-11", entity.taxCode());
    }

    @Test
    public void testJacksonTypeTagWithoutSelectedDiscriminatorCreatesExplicitBranch() throws Exception {
        JsonSubTypes jsonSubTypes = ClientPatchInput.class.getAnnotation(JsonSubTypes.class);
        JsonTypeName jsonTypeName = ClientPatchInput.Organization.class.getAnnotation(JsonTypeName.class);

        Assertions.assertNotNull(jsonSubTypes);
        Assertions.assertEquals(1, jsonSubTypes.value().length);
        Assertions.assertEquals(ClientPatchInput.Organization.class, jsonSubTypes.value()[0].value());
        Assertions.assertEquals("", jsonSubTypes.value()[0].name());
        Assertions.assertNotNull(jsonTypeName);
        Assertions.assertEquals("ORG", jsonTypeName.value());

        ClientPatchInput input = mapper.readValue(
                "{\"type\":\"ORG\",\"id\":21,\"name\":\"Org json\",\"taxCode\":\"T-21\"}",
                ClientPatchInput.class
        );

        Assertions.assertTrue(input instanceof ClientPatchInput.Organization);

        Organization entity = ((ClientPatchInput.Organization) input).toEntity();
        Assertions.assertEquals(Organization.class, ((ImmutableSpi) entity).__type().getJavaClass());
        Assertions.assertEquals(21L, entity.id());
        Assertions.assertEquals("Org json", entity.name());
        Assertions.assertEquals("T-21", entity.taxCode());
    }

    @Test
    public void testJacksonMissingTypeTagWithoutSelectedDiscriminatorUsesDefaultBranch() throws Exception {
        ClientPatchInput input = mapper.readValue(
                "{\"id\":22,\"name\":\"Default json\"}",
                ClientPatchInput.class
        );

        Assertions.assertTrue(input instanceof ClientPatchInput.Default);

        Client entity = input.toEntity();
        Assertions.assertEquals(Client.class, ((ImmutableSpi) entity).__type().getJavaClass());
        Assertions.assertEquals(22L, entity.id());
        Assertions.assertEquals("Default json", entity.name());
    }

    @Test
    public void testDefaultInputWithDiscriminatorCreatesDerivedTypeEntityShape() {
        ClientDiscriminatorInput.Default input = new ClientDiscriminatorInput.Default();
        input.setId(12L);
        input.setType("Person");
        input.setName("Person patch");

        Client entity = input.toEntity();

        Assertions.assertEquals(Person.class, ((ImmutableSpi) entity).__type().getJavaClass());
        Assertions.assertEquals(12L, entity.id());
        Assertions.assertEquals("Person patch", entity.name());
    }

    @Test
    public void testJacksonUnknownTypeTagWithSelectedDiscriminatorUsesDefaultBranchAndKeepsValue() throws Exception {
        ClientDiscriminatorInput input = mapper.readValue(
                "{\"type\":\"Person\",\"id\":23,\"name\":\"Person json\"}",
                ClientDiscriminatorInput.class
        );

        Assertions.assertTrue(input instanceof ClientDiscriminatorInput.Default);
        Assertions.assertEquals("Person", input.getType());

        Client entity = input.toEntity();
        Assertions.assertEquals(Person.class, ((ImmutableSpi) entity).__type().getJavaClass());
        Assertions.assertEquals(23L, entity.id());
        Assertions.assertEquals("Person json", entity.name());
    }

    @Test
    public void testJacksonSelectedDiscriminatorUsesDtoAlias() throws Exception {
        ClientAliasedDiscriminatorInput input = mapper.readValue(
                "{\"kind\":\"ORG\",\"id\":24,\"name\":\"Aliased org\",\"taxCode\":\"T-24\"}",
                ClientAliasedDiscriminatorInput.class
        );

        Assertions.assertTrue(input instanceof ClientAliasedDiscriminatorInput.Organization);
        Assertions.assertEquals("ORG", input.getKind());

        Organization entity = ((ClientAliasedDiscriminatorInput.Organization) input).toEntity();
        Assertions.assertEquals(Organization.class, ((ImmutableSpi) entity).__type().getJavaClass());
        Assertions.assertEquals(24L, entity.id());
        Assertions.assertEquals("Aliased org", entity.name());
        Assertions.assertEquals("T-24", entity.taxCode());
    }

    @Test
    public void testUserJsonTypeInfoIsNotOverridden() throws Exception {
        JsonTypeInfo jsonTypeInfo = ClientCustomJsonTypeInfoInput.class.getAnnotation(JsonTypeInfo.class);

        Assertions.assertNotNull(jsonTypeInfo);
        Assertions.assertEquals("customType", jsonTypeInfo.property());

        ClientCustomJsonTypeInfoInput input = mapper.readValue(
                "{\"customType\":\"ORG\",\"id\":25,\"name\":\"Custom type info\",\"taxCode\":\"T-25\"}",
                ClientCustomJsonTypeInfoInput.class
        );

        Assertions.assertTrue(input instanceof ClientCustomJsonTypeInfoInput.Organization);
    }

    @Test
    public void testUserJsonTypeNameIsNotOverridden() throws Exception {
        JsonTypeName jsonTypeName =
                ClientCustomJsonTypeNameInput.Organization.class.getAnnotation(JsonTypeName.class);

        Assertions.assertNotNull(jsonTypeName);
        Assertions.assertEquals("company", jsonTypeName.value());

        ClientCustomJsonTypeNameInput input = mapper.readValue(
                "{\"type\":\"company\",\"id\":26,\"name\":\"Custom type name\",\"taxCode\":\"T-26\"}",
                ClientCustomJsonTypeNameInput.class
        );

        Assertions.assertTrue(input instanceof ClientCustomJsonTypeNameInput.Organization);
    }

    @Test
    public void testUserJsonSubTypesOwnsBranchNames() {
        JsonSubTypes jsonSubTypes = ClientCustomJsonSubTypesInput.class.getAnnotation(JsonSubTypes.class);
        JsonTypeName jsonTypeName =
                ClientCustomJsonSubTypesInput.Organization.class.getAnnotation(JsonTypeName.class);

        Assertions.assertNotNull(jsonSubTypes);
        Assertions.assertEquals(1, jsonSubTypes.value().length);
        Assertions.assertEquals(String.class, jsonSubTypes.value()[0].value());
        Assertions.assertNull(jsonTypeName);
    }

    @Test
    public void testBranchAndDiscriminatorInputCanAgree() {
        ClientDiscriminatorInput.Organization input = new ClientDiscriminatorInput.Organization();
        input.setId(13L);
        input.setType("ORG");
        input.setName("Org patch");
        input.setTaxCode("T-13");

        Organization entity = input.toEntity();

        Assertions.assertEquals(Organization.class, ((ImmutableSpi) entity).__type().getJavaClass());
        Assertions.assertEquals(13L, entity.id());
        Assertions.assertEquals("Org patch", entity.name());
        Assertions.assertEquals("T-13", entity.taxCode());
    }

    @Test
    public void testBranchAndDiscriminatorInputMismatchIsRejected() {
        ClientDiscriminatorInput.Organization input = new ClientDiscriminatorInput.Organization();
        input.setId(13L);
        input.setType("Person");
        input.setName("Org patch");
        input.setTaxCode("T-13");

        IllegalArgumentException ex = Assertions.assertThrows(
                IllegalArgumentException.class,
                input::toEntity
        );

        Assertions.assertEquals(
                "Discriminator value \"Person\" does not match polymorphic input DTO branch " +
                        "\"org.babyfish.jimmer.sql.model.inheritance.singletable.dto." +
                        "ClientDiscriminatorInput.Organization\" whose entity type is " +
                        "\"org.babyfish.jimmer.sql.model.inheritance.singletable.Organization\"",
                ex.getMessage()
        );
    }

    @Test
    public void testEnumDefaultInputWithRootDiscriminatorCreatesRootEntityShape() {
        EnumClientDiscriminatorInput.Default input = new EnumClientDiscriminatorInput.Default();
        input.setId(130L);
        input.setType(ClientType.CLIENT);
        input.setName("Enum client");

        EnumClient entity = input.toEntity();

        Assertions.assertEquals(EnumClient.class, ((ImmutableSpi) entity).__type().getJavaClass());
        Assertions.assertEquals(130L, entity.id());
        Assertions.assertEquals("Enum client", entity.name());
    }

    @Test
    public void testEnumDefaultInputWithDerivedTypeDiscriminatorCreatesDerivedTypeEntityShape() {
        EnumClientDiscriminatorInput.Default input = new EnumClientDiscriminatorInput.Default();
        input.setId(131L);
        input.setType(ClientType.PERSON);
        input.setName("Enum person");

        EnumClient entity = input.toEntity();

        Assertions.assertEquals(EnumPerson.class, ((ImmutableSpi) entity).__type().getJavaClass());
        Assertions.assertEquals(131L, entity.id());
        Assertions.assertEquals("Enum person", entity.name());
    }

    @Test
    public void testEnumBranchAndDiscriminatorInputMismatchIsRejected() {
        EnumClientDiscriminatorInput.Organization input = new EnumClientDiscriminatorInput.Organization();
        input.setId(132L);
        input.setType(ClientType.PERSON);
        input.setName("Enum org");

        IllegalArgumentException ex = Assertions.assertThrows(
                IllegalArgumentException.class,
                input::toEntity
        );

        Assertions.assertEquals(
                "Discriminator value \"PERSON\" does not match polymorphic input DTO branch " +
                        "\"org.babyfish.jimmer.sql.model.inheritance.enumdiscriminator.dto." +
                        "EnumClientDiscriminatorInput.Organization\" whose entity type is " +
                        "\"org.babyfish.jimmer.sql.model.inheritance.enumdiscriminator.EnumOrganization\"",
                ex.getMessage()
        );
    }

    @Test
    public void testExhaustiveGeneratedDerivedTypeInputCreatesDerivedTypeEntityShape() {
        ClientExhaustiveInput.Person input = new ClientExhaustiveInput.Person();
        input.setId(14L);
        input.setName("Person patch");
        input.setFirstName("Ann");
        input.setLastName("Smith");

        Person entity = input.toEntity();

        Assertions.assertEquals(Person.class, ((ImmutableSpi) entity).__type().getJavaClass());
        Assertions.assertEquals(14L, entity.id());
        Assertions.assertEquals("Person patch", entity.name());
        Assertions.assertEquals("Ann", entity.firstName());
        Assertions.assertEquals("Smith", entity.lastName());
    }

    @Test
    public void testInstantiableRootDefaultInputCreatesRootEntityShape() {
        InstantiableClientDefaultInput.Base input = new InstantiableClientDefaultInput.Base();
        input.setId(15L);
        input.setName("Joined root patch");

        org.babyfish.jimmer.sql.model.inheritance.joinedtable.instantiable.Client entity =
                input.toEntity();

        Assertions.assertEquals(
                org.babyfish.jimmer.sql.model.inheritance.joinedtable.instantiable.Client.class,
                ((ImmutableSpi) entity).__type().getJavaClass()
        );
        Assertions.assertEquals(15L, entity.id());
        Assertions.assertEquals("Joined root patch", entity.name());
    }
}

File: project/jimmer-sql-kotlin/src/test/kotlin/org/babyfish/jimmer/sql/kt/dto/PolymorphicDtoInputTest.kt
package org.babyfish.jimmer.sql.kt.dto

import com.fasterxml.jackson.annotation.JsonSubTypes
import com.fasterxml.jackson.annotation.JsonTypeInfo
import com.fasterxml.jackson.annotation.JsonTypeName
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import com.fasterxml.jackson.module.kotlin.registerKotlinModule
import org.babyfish.jimmer.runtime.ImmutableSpi
import org.babyfish.jimmer.sql.kt.model.inheritance.enumdiscriminator.KClientType
import org.babyfish.jimmer.sql.kt.model.inheritance.enumdiscriminator.KEnumClient
import org.babyfish.jimmer.sql.kt.model.inheritance.enumdiscriminator.KEnumPerson
import org.babyfish.jimmer.sql.kt.model.inheritance.enumdiscriminator.dto.KEnumClientDiscriminatorInput
import org.babyfish.jimmer.sql.kt.model.inheritance.joinedtable.instantiable.dto.KInstantiableClientDefaultInput
import org.babyfish.jimmer.sql.kt.model.inheritance.singletable.KClient
import org.babyfish.jimmer.sql.kt.model.inheritance.singletable.KOrganization
import org.babyfish.jimmer.sql.kt.model.inheritance.singletable.KPerson
import org.babyfish.jimmer.sql.kt.model.inheritance.singletable.dto.*
import org.babyfish.jimmer.sql.kt.model.inheritance.singletable.dto.KClientCustomJsonSubTypesInput
import kotlin.test.*
import org.babyfish.jimmer.sql.kt.model.inheritance.joinedtable.instantiable.KClient as KInstantiableClient

class PolymorphicDtoInputTest {

    private val mapper = ObjectMapper().registerKotlinModule()

    @Test
    fun testImplicitDefaultInputCreatesRootEntityShape() {
        val entity = KClientPatchInput.Default(
            id = 10L,
            name = "Base patch"
        ).toEntity()

        assertEquals(KClient::class.java, (entity as ImmutableSpi).__type().javaClass)
        assertEquals(10L, entity.id)
        assertEquals("Base patch", entity.name)
    }

    @Test
    fun testExplicitBranchInputCreatesDerivedTypeEntityShape() {
        val entity = KClientPatchInput.Organization(
            id = 11L,
            name = "Org patch",
            taxCode = "T-11"
        ).toEntity()

        assertEquals(KOrganization::class.java, (entity as ImmutableSpi).__type().javaClass)
        assertEquals(11L, entity.id)
        assertEquals("Org patch", entity.name)
        assertEquals("T-11", entity.taxCode)
    }

    @Test
    fun testJacksonTypeTagWithoutSelectedDiscriminatorCreatesExplicitBranch() {
        val jsonSubTypes = KClientPatchInput::class.java.getAnnotation(JsonSubTypes::class.java)
        val jsonTypeName = KClientPatchInput.Organization::class.java.getAnnotation(JsonTypeName::class.java)

        assertNotNull(jsonSubTypes)
        assertEquals(1, jsonSubTypes.value.size)
        assertEquals(KClientPatchInput.Organization::class.java, jsonSubTypes.value[0].value.java)
        assertEquals("", jsonSubTypes.value[0].name)
        assertNotNull(jsonTypeName)
        assertEquals("ORG", jsonTypeName.value)

        val input = mapper.readValue<KClientPatchInput>(
            """{"type":"ORG","id":21,"name":"Org json","taxCode":"T-21"}"""
        )

        assertEquals(KClientPatchInput.Organization::class, input::class)

        val entity = (input as KClientPatchInput.Organization).toEntity()
        assertEquals(KOrganization::class.java, (entity as ImmutableSpi).__type().javaClass)
        assertEquals(21L, entity.id)
        assertEquals("Org json", entity.name)
        assertEquals("T-21", entity.taxCode)
    }

    @Test
    fun testJacksonSelectedDiscriminatorUsesDtoAlias() {
        val input = mapper.readValue<KClientAliasedDiscriminatorInput>(
            """{"kind":"ORG","id":22,"name":"Aliased org","taxCode":"T-22"}"""
        )

        assertEquals(KClientAliasedDiscriminatorInput.Organization::class, input::class)
        assertEquals("ORG", input.kind)

        val entity = (input as KClientAliasedDiscriminatorInput.Organization).toEntity()
        assertEquals(KOrganization::class.java, (entity as ImmutableSpi).__type().javaClass)
        assertEquals(22L, entity.id)
        assertEquals("Aliased org", entity.name)
        assertEquals("T-22", entity.taxCode)
    }

    @Test
    fun testUserJsonTypeInfoIsNotOverridden() {
        val jsonTypeInfo = KClientCustomJsonTypeInfoInput::class.java.getAnnotation(JsonTypeInfo::class.java)

        assertNotNull(jsonTypeInfo)
        assertEquals("customType", jsonTypeInfo.property)

        val input = mapper.readValue<KClientCustomJsonTypeInfoInput>(
            """{"customType":"ORG","id":23,"name":"Custom type info","taxCode":"T-23"}"""
        )

        assertEquals(KClientCustomJsonTypeInfoInput.Organization::class, input::class)
    }

    @Test
    fun testUserJsonTypeNameIsNotOverridden() {
        val jsonTypeName = KClientCustomJsonTypeNameInput.Organization::class.java
            .getAnnotation(JsonTypeName::class.java)

        assertNotNull(jsonTypeName)
        assertEquals("company", jsonTypeName.value)

        val input = mapper.readValue<KClientCustomJsonTypeNameInput>(
            """{"type":"company","id":24,"name":"Custom type name","taxCode":"T-24"}"""
        )

        assertEquals(KClientCustomJsonTypeNameInput.Organization::class, input::class)
    }

    @Test
    fun testUserJsonSubTypesOwnsBranchNames() {
        val jsonSubTypes = KClientCustomJsonSubTypesInput::class.java.getAnnotation(JsonSubTypes::class.java)
        val jsonTypeName = KClientCustomJsonSubTypesInput.Organization::class.java
            .getAnnotation(JsonTypeName::class.java)

        assertNotNull(jsonSubTypes)
        assertEquals(1, jsonSubTypes.value.size)
        assertEquals(String::class.java, jsonSubTypes.value[0].value.java)
        assertNull(jsonTypeName)
    }

    @Test
    fun testDefaultInputWithDiscriminatorCreatesDerivedTypeEntityShape() {
        val entity = KClientDiscriminatorInput.Default(
            id = 12L,
            type = "KPerson",
            name = "Person patch"
        ).toEntity()

        assertEquals(KPerson::class.java, (entity as ImmutableSpi).__type().javaClass)
        assertEquals(12L, entity.id)
        assertEquals("Person patch", entity.name)
    }

    @Test
    fun testBranchAndDiscriminatorInputCanAgree() {
        val entity = KClientDiscriminatorInput.Organization(
            id = 13L,
            type = "ORG",
            name = "Org patch",
            taxCode = "T-13"
        ).toEntity()

        assertEquals(KOrganization::class.java, (entity as ImmutableSpi).__type().javaClass)
        assertEquals(13L, entity.id)
        assertEquals("Org patch", entity.name)
        assertEquals("T-13", entity.taxCode)
    }

    @Test
    fun testBranchAndDiscriminatorInputMismatchIsRejected() {
        val ex = assertFailsWith<IllegalArgumentException> {
            KClientDiscriminatorInput.Organization(
                id = 13L,
                type = "KPerson",
                name = "Org patch",
                taxCode = "T-13"
            ).toEntity()
        }

        assertEquals(
            "Discriminator value \"KPerson\" does not match polymorphic input DTO branch " +
                    "\"org.babyfish.jimmer.sql.kt.model.inheritance.singletable.dto." +
                    "KClientDiscriminatorInput.Organization\" whose entity type is " +
                    "\"org.babyfish.jimmer.sql.kt.model.inheritance.singletable.KOrganization\"",
            ex.message
        )
    }

    @Test
    fun testEnumDefaultInputWithRootDiscriminatorCreatesRootEntityShape() {
        val entity = KEnumClientDiscriminatorInput.Default(
            id = 130L,
            type = KClientType.CLIENT,
            name = "Enum client"
        ).toEntity()

        assertEquals(KEnumClient::class.java, (entity as ImmutableSpi).__type().javaClass)
        assertEquals(130L, entity.id)
        assertEquals("Enum client", entity.name)
    }

    @Test
    fun testEnumDefaultInputWithDerivedTypeDiscriminatorCreatesDerivedTypeEntityShape() {
        val entity = KEnumClientDiscriminatorInput.Default(
            id = 131L,
            type = KClientType.PERSON,
            name = "Enum person"
        ).toEntity()

        assertEquals(KEnumPerson::class.java, (entity as ImmutableSpi).__type().javaClass)
        assertEquals(131L, entity.id)
        assertEquals("Enum person", entity.name)
    }

    @Test
    fun testEnumBranchAndDiscriminatorInputMismatchIsRejected() {
        val ex = assertFailsWith<IllegalArgumentException> {
            KEnumClientDiscriminatorInput.Organization(
                id = 132L,
                type = KClientType.PERSON,
                name = "Enum org"
            ).toEntity()
        }

        assertEquals(
            "Discriminator value \"PERSON\" does not match polymorphic input DTO branch " +
                    "\"org.babyfish.jimmer.sql.kt.model.inheritance.enumdiscriminator.dto." +
                    "KEnumClientDiscriminatorInput.Organization\" whose entity type is " +
                    "\"org.babyfish.jimmer.sql.kt.model.inheritance.enumdiscriminator.KEnumOrganization\"",
            ex.message
        )
    }

    @Test
    fun testExhaustiveGeneratedDerivedTypeInputCreatesDerivedTypeEntityShape() {
        val entity = KClientExhaustiveInput.Person(
            id = 14L,
            name = "Person patch",
            firstName = "Ann",
            lastName = "Smith"
        ).toEntity()

        assertEquals(KPerson::class.java, (entity as ImmutableSpi).__type().javaClass)
        assertEquals(14L, entity.id)
        assertEquals("Person patch", entity.name)
        assertEquals("Ann", entity.firstName)
        assertEquals("Smith", entity.lastName)
    }

    @Test
    fun testInstantiableRootDefaultInputCreatesRootEntityShape() {
        val entity = KInstantiableClientDefaultInput.Base(
            id = 15L,
            name = "Joined root patch"
        ).toEntity()

        assertEquals(KInstantiableClient::class.java, (entity as ImmutableSpi).__type().javaClass)
        assertEquals(15L, entity.id)
        assertEquals("Joined root patch", entity.name)
    }
}
